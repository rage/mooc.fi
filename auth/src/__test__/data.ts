import { NextFunction, Request, Response } from "express"

export const testProfile = {
  edu_person_affiliation: "",
  edu_person_principal_name: "test",
  schac_home_organization: "helsinki.fi",
  organizational_unit: "organizational_unit",
  schac_personal_unique_code:
    "urn:schac:personalUniqueCode:int:studentID:helsinki.fi:121345678",
  display_name: "test",
  given_name: "test",
  mail: "test@mail.com",
}

export const req = {
  body: {},
  query: {
    RelayState: JSON.stringify({
      language: "fi",
      provider: "hy",
      action: "sign-in",
    }),
  },
  params: {},
  login: jest.fn().mockImplementation((_: any, fn: Function) => fn(null)),
  logout: jest.fn(),
} as unknown as Request
export const res = {
  redirect: jest.fn().mockReturnThis(),
  locals: {},
  setMOOCCookies: jest.fn().mockReturnThis(),
} as unknown as Response
export const next = jest.fn() as NextFunction

export const okTokenResponse = {
  data: {
    status: 200,
    success: true,
    tmc_token: "tmc_token",
    access_token: "access_token",
    admin: false,
  },
}
