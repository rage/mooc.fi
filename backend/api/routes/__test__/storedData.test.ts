import {
  createRequestHelpers,
  FAKE_NORMAL_USER_AUTHORIZATION_HEADERS,
  getTestContext,
  RequestGet,
  RequestPost,
  setupTMCWithDefaultFakeUsers,
} from "../../../tests"
import { seed } from "../../../tests/data"

const ctx = getTestContext()

describe("API", () => {
  describe("/api/temporary-stored-data", () => {
    setupTMCWithDefaultFakeUsers()

    let getStoredData: (slug: string) => RequestGet
    let postStoredData: (slug: string) => RequestPost

    beforeEach(async () => {
      await seed(ctx.prisma)
      const { get, post } = createRequestHelpers(ctx.port)
      getStoredData = (slug: string) =>
        get(`/api/temporary-stored-data/${slug}`, {})
      postStoredData = (slug: string) =>
        post(`/api/temporary-stored-data/${slug}`, {})
    })

    describe("post", () => {
      it("errors on no auth", async () => {
        return postStoredData("course1")({
          data: { data: "foo" },
        })
          .then(() => fail())
          .catch(({ response }) => {
            expect(response.status).toBe(401)
          })
      })

      it("errors on no data", async () => {
        return postStoredData("course1")({
          headers: FAKE_NORMAL_USER_AUTHORIZATION_HEADERS,
        })
          .then(() => fail())
          .catch(({ response }) => {
            expect(response.status).toBe(400)
          })
      })

      it("errors on invalid slug", async () => {
        return postStoredData("foobarbaz")({
          headers: FAKE_NORMAL_USER_AUTHORIZATION_HEADERS,
          data: { data: "foo" },
        })
          .then(() => fail())
          .catch(({ response }) => {
            expect(response.status).toBe(401)
            expect(response.data.error).toContain(
              "course with slug or course alias with course code foobarbaz",
            )
          })
      })

      it("creates stored data", async () => {
        const existing = await ctx.prisma.storedData.findFirst({
          where: {
            user_id: "20000000000000000000000000000102",
            course_id: "00000000000000000000000000000002",
          },
        })

        expect(existing).toBeNull()

        const res = await postStoredData("course1")({
          data: { data: "foo foo" },
          headers: FAKE_NORMAL_USER_AUTHORIZATION_HEADERS,
        })

        expect(res.status).toBe(200)
        expect(res.data.message).toContain("stored data created")

        const created = await ctx.prisma.storedData.findFirst({
          where: {
            user_id: "20000000000000000000000000000102",
            course_id: "00000000000000000000000000000002",
          },
        })

        expect(created?.data).toEqual("foo foo")
      })

      it("updates stored data", async () => {
        const existing = await ctx.prisma.storedData.findFirst({
          where: {
            user_id: "20000000000000000000000000000102",
            course_id: "00000000000000000000000000000001",
          },
        })

        expect(existing?.data).toEqual("user1_foo")

        const res = await postStoredData("course2")({
          data: { data: "foo foo" },
          headers: FAKE_NORMAL_USER_AUTHORIZATION_HEADERS,
        })

        expect(res.status).toBe(200)
        expect(res.data.message).toEqual("stored data updated")

        const updated = await ctx.prisma.storedData.findFirst({
          where: {
            user_id: "20000000000000000000000000000102",
            course_id: "00000000000000000000000000000001",
          },
        })

        expect(updated?.data).toEqual("foo foo")
      })
    })

    describe("get", () => {
      it("errors on no auth", async () => {
        return getStoredData("course1")({})
          .then(() => fail())
          .catch(({ response }) => {
            expect(response.status).toBe(401)
          })
      })

      it("errors on invalid slug", async () => {
        return getStoredData("foobarbaz")({
          headers: FAKE_NORMAL_USER_AUTHORIZATION_HEADERS,
        })
          .then(() => fail())
          .catch(({ response }) => {
            expect(response.status).toBe(401)
            expect(response.data.error).toContain("course with slug foobarbaz")
          })
      })

      it("errors on non-owned course", async () => {
        return getStoredData("course1")({
          headers: FAKE_NORMAL_USER_AUTHORIZATION_HEADERS,
        })
          .then(() => fail())
          .catch(({ response }) => {
            expect(response.status).toBe(401)
            expect(response.data.message).toContain("no ownership")
          })
      })

      it("returns stored data from owned course", async () => {
        const res = await getStoredData("course2")({
          headers: FAKE_NORMAL_USER_AUTHORIZATION_HEADERS,
        })

        expect(res.data).toMatchSnapshot()
      })
    })
  })
})
