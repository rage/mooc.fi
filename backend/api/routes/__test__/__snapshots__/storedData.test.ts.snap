// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API /api/temporary-stored-data get returns stored data from owned course 1`] = `
[
  {
    "completions": [
      {
        "certificate_id": null,
        "completion_date": "2000-01-01T08:00:00.000Z",
        "completion_language": "en",
        "completion_registration_attempt_date": "2022-01-01T08:00:00.000Z",
        "course_id": "00000000-0000-0000-0000-000000000001",
        "created_at": "2000-01-01T08:00:00.000Z",
        "eligible_for_ects": true,
        "email": "e@mail.com",
        "grade": null,
        "id": "30000000-0000-0000-0000-000000000103",
        "student_number": null,
        "tier": null,
        "updated_at": "2000-01-01T08:00:00.000Z",
        "user_id": "20000000-0000-0000-0000-000000000102",
        "user_upstream_id": 1,
      },
    ],
    "storedData": {
      "course_id": "00000000-0000-0000-0000-000000000001",
      "created_at": "1900-01-01T08:00:00.000Z",
      "data": "user1_foo",
      "updated_at": "1900-01-01T08:00:00.000Z",
      "user_id": "20000000-0000-0000-0000-000000000102",
    },
    "user": {
      "administrator": false,
      "created_at": "1900-01-01T08:00:00.000Z",
      "email": "e@mail.com",
      "first_name": null,
      "id": "20000000-0000-0000-0000-000000000102",
      "last_name": null,
      "real_student_number": null,
      "research_consent": null,
      "student_number": null,
      "updated_at": "1900-01-01T08:00:00.000Z",
      "upstream_id": 1,
      "username": "existing_user",
    },
  },
]
`;
